Version locally the script geany-release.py so far hosted in the wiki…

giuspen opened the pull request:
- version locally the script `geany-release.py` so far hosted in the wiki
- applied fix for building without signing (missing key/certificate in the filesystem)
- added copying geany.nsi from the build directory into the root directory so that no need to run configure twice
- changed version num to 2.1

b4n commented on scripts/geany-release.py:
```diff
@@ -0,0 +1,135 @@
+SIGN_CERTIFICATE_KEY = join(BASE_DIR, 'codesign_key.pem')  # adjust to your needs
+SIGN_WEBSITE = 'https://www.geany.org'
+SIGN_NAME = 'Geany Binary'
+SIGN_TIMESTAMP = 'https://zeitstempel.dfn.de/'
```
After a bit of research this looks probably OK, but without further explanation it looks *so* shady it'd be good to explain it.  @eht16 can probably help here(?)

eht16 replied:
I explained parts of the signing topic in https://github.com/geany/geany/issues/4189#issuecomment-2629001461.

The short story is, we will probably not sign the binaries any longer and so we could also remove the related code.

eht16 replied:
@giuspen since we will not use this kind of code signing any longer, feel free to remove the relevant code.

b4n commented:
If this file is in the repository, its name should reflect that it's for *Windows* releases, not the usual source releases (which is covered by *doc/making-a-release*).

giuspen commented:
Makes sense @b4n I will rename to msys2-geany-release.py if that's ok
Forgot to mention: the original script from @eht16  is https://wiki.geany.org/_media/howtos/win32/geany-release.py.txt linked from https://wiki.geany.org/howtos/win32/msys2

giuspen commented:
I have modified the native windows installer script to support running in the native windows workflow and added the workflow to build the installer on a windows native builder.
This for now is only triggered manually.

eht16 commented on .github/workflows/windows-msys2-build.yml:
```diff
@@ -2,7 +2,7 @@ name: Windows/MSYS2 Build
 on: [workflow_dispatch]
 
 jobs:
-  msys2-ucrt64:
```
Just for my curiosity, did the `ucrt64` variant cause problems when building?

giuspen replied:
No I simply noticed that you are fetching the mingw64 packages in your dependency bundle bash scripts so didn't want to do too many things at once.
Since it now works, I can apply new changes for the ucrt64 but that will involve changing your other script.

eht16 replied:
Ah I see, alright.
I personally have no preference on which variant to use. I think I used `mingw64` only because I always used it :D.

giuspen replied:
I read that ucrt64 has better performance so I think we should migrate, but I would like before to complete also geany-plugins where I am struggling with an Autotools issue not reproducible on my Windows ( https://github.com/giuspen/geany-plugins/actions/runs/15654226235/job/44103094132 )

giuspen replied:
I did try multiple solutions with the newlines but got no progress, then Google Gemini suggested me some changes which and I made some progress, other changes still necessary ( https://github.com/giuspen/geany-plugins/actions/runs/15662106128/job/44121339614 ) I will try and sort this out before going to sleep and update https://github.com/geany/geany-plugins/pull/1408

giuspen replied:
You were right about the newlines Enrico, I must have done something wrong at the beginning when converting... anyway now I got that sorted with dos2unix calls from the workflow action

eht16 commented:
> I have modified the native windows installer script to support running in the native windows workflow

Nice, I didn't expect it would be that easy (at least from reading the result, I guess the way to a working state was not that easy :D).

eht16 commented on scripts/msys2-geany-release.py:
```diff
@@ -0,0 +1,122 @@
+"""
+This script prepares a Geany release on Windows.
+The following steps will be executed:
+- update geany-themes repository
```
Not strictly related to this PR but I don't want to get it missed before the release:

Thanks to #4044, we don't need to manually pull the geany-themes files any longer as (most) they are already included in Geany.

If you like, just drop the relevant code completely.

giuspen replied:
Ok I will apply the change later today or tomorrow

giuspen replied:
done

giuspen commented:
> > I have modified the native windows installer script to support running in the native windows workflow
> 
> Nice, I didn't expect it would be that easy (at least from reading the result, I guess the way to a working state was not that easy :D).

It's fairly simple if the builder doesn't present different behaviour to my local machine... which unfortunately is happening on the geany-plugins 😖

b4n commented on .github/workflows/windows-msys2-build.yml:
```diff
@@ -0,0 +1,57 @@
+      - name: CI-Build
+        run: |
+          DESTINATON=$(cygpath -u "${GITHUB_WORKSPACE}")/geany_build
+          VERSION=$(grep 'AC_INIT' configure.ac | sed -E 's/AC_INIT\(\[[^]]+\],\s*\[([^]]+)\].*/\1/')
```
If you can afford it, a probably safer solution is to call autom4te:
`autom4te --no-cache --language=Autoconf-without-aclocal-m4 --trace AC_INIT:\$2 configure.ac`

Otherwise, let's make it a slightly more forgiving sed-only call:
`sed -rn 's/^AC_INIT[(][^,]*,\s*\[?([^],]*)\]?\s*,.*$/\1/p' configure.ac`

giuspen replied:
applied suggested improvement, thanks

giuspen commented on scripts/msys2-geany-release.py:
```diff
@@ -0,0 +1,122 @@
+                print(f"GOT VERSION {VERSION} FROM {line}")
+            else:
+                print(f"!! FAILED TO GET VERSION FROM {line}")
+            break # Stop after finding the first AC_INIT line with a version
```
applied suggested improvement, thanks

b4n commented on scripts/msys2-geany-release.py:
```diff
@@ -0,0 +1,99 @@
+with open(join(SOURCE_DIR, 'configure.ac'), 'r') as f:
+    ver = next((l.split(',')[1].strip(' []') for l in f if l.startswith('AC_INIT')), None)
+    if ver is None:
+        print(f"!! FAILED TO GET VERSION FROM {line}")
```
Sorry, m'y initial suggestion had an issue, and you probably didn't see the edit:
```suggestion
        print(f"!! FAILED TO GET VERSION FROM {f.name}")
```

giuspen commented:
This is now complete although this only produces the installer for Geany. Similar PR on the plugins produces both https://github.com/geany/geany-plugins/pull/1408
so I'm tempted to use that in order to generate the signed installers.
Another option would be to build here also the plugins instead. Preferences?
Another topic is the migration to UCRT64 as that is advised from msys2 but my concern is that the release is very close and we are running out of time, so I'm leaning towards doing that just after the 2.1 release.
Mainly, as I need to write to SignPath, I need to indicate the workflow that we intend to use to do the signing and that has to be on the official repo, not my fork, so I will hold fire until my PRs are merged.

b4n commented:
Remembering I don't know anything about the subject, but:

> This is now complete although this only produces the installer for Geany. Similar PR on the plugins produces both [geany/geany-plugins#1408](https://github.com/geany/geany-plugins/pull/1408) so I'm tempted to use that in order to generate the signed installers. Another option would be to build here also the plugins instead. Preferences?

Ideally each installer would be independent -- especially as we usually release Geany first, and only a little bit after plugins.
If it's not doable (for now?), I guess doing it in GP is better because when we release Geany we don't necessarily have a ready GP.

> Another topic is the migration to UCRT64 as that is advised from msys2 but my concern is that the release is very close and we are running out of time, so I'm leaning towards doing that just after the 2.1 release.

I guess that's wise indeed, just play it safe here.

> Mainly, as I need to write to SignPath, I need to indicate the workflow that we intend to use to do the signing and that has to be on the official repo, not my fork, so I will hold fire until my PRs are merged.

Yeah so we need to get this sorted out if we want a chance to have something in less than 3 weeks…

eht16 commented:
>We can ask for signatures on two different repos / two workflows, one
>for Geany and one for Geany-Plugins, I don't see reason why they won't
>give us both (I didn't know that the two are released separately).

Sounds good.
I think if possible it is better when we keep Geany and G-P independent.

For the UCRT64 migration: yeah, let's postpone this to after 2.1. I
hear a nice, friendly voice in my mind warning about last minute changes
before a release :).

I didn't follow the changes in detail yet but I'd say maybe squash all
the commits together or to just a few commits and then we could merge
it soon. If it's not yet alright, we can still improve it after it has
been merged, as long as the SignPath process can keep going.

Same goes for https://github.com/geany/geany-plugins/pull/1408.

giuspen commented on .gitignore:
```diff
@@ -22,6 +22,7 @@ Makefile.in
 /aclocal.m4
 /autom4te.cache
 /build/
+/_build/
```
In the Windows build we use that directory to build, see https://github.com/giuspen/geany/blob/0ef59cca4d7f927c828bd41ca9460e677f03b632/.github/workflows/windows-msys2-build.yml#L39 and also in the wiki guide of building for a release https://wiki.geany.org/howtos/win32/msys2

b4n replied:
Yeah I've seen it's used in the CI, but I doubted the builder cared :)
But fair enough if it's the recommended name for manual builds as well, makes sense.  And FWIW it's a name I usually use, so I won't complain :)

b4n commented:
@eht16 if we need to merge this for the Windows release, maybe we should get on with it?  I don't mean to push but it's getting closer my the minute :wink:  If not, we can postpone, but IIUC if we want to use it for the next release we'll have to be ready a bit ahead of time.  Same goes for https://github.com/geany/geany-plugins/pull/1408

You call though, I don't know nearly enough about any of this to give a meaningful opinion on the changes themselves.

eht16 commented:
Yeah, content-wise it's totally fine for me, as said before.

I think after all, it's only about the commit history.

@giuspen do you think you can clean it up soonish? Otherwise I could give it a try.

Same goes for https://github.com/geany/geany-plugins/pull/1408.

giuspen commented:
Sorry I thought you were happy to squash and merge via the GitHub button. I can certainly create 2 new pull requests with a single commit tomorrow, 2 new branches.

eht16 commented:
Let's keep it simple, for me it's fine to squash it all into one commit.

giuspen commented:
Thanks @b4n I will try and write to SignPath later this evening.

b4n commented:
@giuspen OK, triggered

eht16 commented:
> @giuspen OK, triggered

And it's finished successfully. Great work!
